File: immiglensBE/app/schemas/organization.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, field_validator

from app.models.organization import OrgRole


class OrganizationCreate(BaseModel):
    name: str

    @field_validator("name")
    @classmethod
    def name_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Organization name cannot be empty")
        return v.strip()


class OrganizationOut(BaseModel):
    id: int
    name: str
    created_by: int
    created_at: datetime

    model_config = {"from_attributes": True}


class OrgMembershipOut(BaseModel):
    id: int
    org_id: int
    user_id: int
    user_name: str = ""
    user_email: str = ""
    role: OrgRole
    joined_at: datetime

    model_config = {"from_attributes": True}


class InviteRequest(BaseModel):
    email: str
    role: OrgRole = OrgRole.VIEWER

    @field_validator("email")
    @classmethod
    def email_valid(cls, v: str) -> str:
        if "@" not in v or "." not in v.split("@")[-1]:
            raise ValueError("Invalid email address")
        return v.strip().lower()


class OrgInvitationOut(BaseModel):
    id: int
    org_id: int
    email: str
    role: OrgRole
    token: str
    expires_at: datetime
    accepted_at: Optional[datetime]
    created_at: datetime

    model_config = {"from_attributes": True}

File: immiglensFE/src/types/organization.ts
export type OrgRole = 'owner' | 'admin' | 'viewer'

export interface Organization {
  id: number
  name: string
  created_by: number
  created_at: string
}

export interface OrgMembership {
  id: number
  org_id: number
  user_id: number
  user_name: string
  user_email: string
  role: OrgRole
  joined_at: string
}

export interface OrgInvitation {
  id: number
  org_id: number
  email: string
  role: OrgRole
  token: string
  expires_at: string
  accepted_at: string | null
  created_at: string
}
